File: Insertion_Sort_List.cpp
#include <iostream>
using namespace std;

struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* findInsertPos(ListNode *head, int x)
{
    ListNode *pre = NULL;
    // if x == INT_MIN, pre will return with NULL if we use only ">"
    for(ListNode *cur = head; cur != NULL && x >= cur->val;
        pre = cur, cur = cur->next)
        ;
    return pre;
}

ListNode* insertionSortList(ListNode *head)
{
    ListNode dummy(INT_MIN);
    for(ListNode *cur = head; cur != NULL;){
        ListNode *pos = findInsertPos(&dummy, cur->val);
        ListNode *tmp = cur->next;
        cur->next = pos->next;
        pos->next = cur;
        cur = tmp;
    }
    return dummy.next;
}


ListNode* generateList(int size)
{
    ListNode *head = (ListNode*)malloc(sizeof(ListNode));
    ListNode *cur = head;
    for(int i = 0; i < size; i++){
    ListNode *p = (ListNode*)malloc(sizeof(ListNode));
    p->val = rand() % 100;
    p->next = NULL;
    cur->next = p;
    cur = cur->next;
    }
    return head->next;
}

void print_list(ListNode *l)
{
    ListNode *tmp = l;
    while(tmp != NULL){
        cout<< tmp->val <<" ";
        tmp = tmp->next;
    }
    cout<<endl;
}

int main()
{
    ListNode *l1 = generateList(20);
    print_list(l1);
    l1 = insertionSortList(l1);
    print_list(l1);
    return 0;
}

File: Insertion_Sort_List.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        p1 = head
        while p1 != None:
            p2 = p1
            min = p2
            
            while p2.next != None:
                if p2.next.val < min.val:                    
                    min = p2.next
                    min_pre = p2
                p2 = p2.next

            min_pre.next = min.next
            min.next = p1.next
            p1.next = min
            p1 = p1.next
        
#main function
sln = Solution()
result = sln.insertionSortList()        
